phoenix-core-client/src/main/java/org/apache/phoenix/compile/ProjectionCompiler.java

```diff
@@ -497,42 +502,106 @@ public static RowProjector compile(StatementContext context, SelectStatement sta
             index++;
         }
 
-        for (int i = arrayProjectedColumnRefs.size() - 1; i >= 0; i--) {
-            Expression expression = arrayProjectedColumnRefs.get(i);
-            Integer count = arrayExpressionCounts.get(expression);
+        for (int i = serverParsedProjectedColumnRefs.size() - 1; i >= 0; i--) {
+            Expression expression = serverParsedProjectedColumnRefs.get(i);
+            Integer count = serverParsedExpressionCounts.get(expression);
             if (count != 0) {
-                arrayKVRefs.remove(i);
-                arrayKVFuncs.remove(i);
-                arrayOldFuncs.remove(i);
+                serverParsedKVRefs.remove(i);
+                serverParsedKVFuncs.remove(i);
+                serverParsedOldFuncs.remove(i);
             }
         }
 
-        if (arrayKVFuncs.size() > 0 && arrayKVRefs.size() > 0) {
-            serailizeArrayIndexInformationAndSetInScan(context, arrayKVFuncs, arrayKVRefs);
+        if (serverParsedKVFuncs.size() > 0 && serverParsedKVRefs.size() > 0) {
```

Comment:
The rest of the code is confusing since JSON functions and ArrayIndex functions are returned in the same output parameter. Can we refactor SelectClauseVisitor and add different getters for ArrayIndex and JSON functions instead of getting them using parameters passed to the constructor?

Reply:
Sorry I didn't understand : SelectClauseVisitor constructor didn't change - which params passed to the constructor are you referring to?
You mean to say instead of using same ServerParsedkvFunctions and Refs - use different variables for arrays and Json and then have different getters for array index and json expressions?. That can be done but these variables are bit spread around for ref counting and other things.

Reply:
The replacement map was only for array index functions (old vs new). Now that JSON path function inherits from the array index function, the code gives the impression that the replacement is needed for the JSON path function too. The code has been introduce to handle this replacement even though there is no such replacement is needed for the JSON functions as far as I understand. That is another reason I suggested that JSON path should not inherit from ArrayIndex but from Scalar. 

The refactoring I suggested is to keep track of array index and JSON functions separately and old vs new replacement should be applied to array index only. Maybe I misunderstood the whole thing.

Reply:
Replacement map is required for JSON functions too. One refactoring I can think of is having a new base class for both ArrayIndexExpression and JSONPathExpression.